<!-- =================================================================================================================================
  - @file     model.original.sdf
  - @author   Krzysztof Pierczyk (krzysztof.pierczyk@gmail.com)
  - @date     Monday, 7th March 2022 3:27:05 pm
  - @modified Monday, 7th March 2022 4:18:56 pm
  - @project  engineering-thesis
  - @brief
  -    
  -    SDF description of the velmobil robot (mobile platform of the WUT Velma robot). Current version is stronly based on the 
  -    source provided by Jakub Sikora (@see @source).
  -    
  - @copyright Krzysztof Pierczyk © 2022
  - @source https://github.com/RCPRG-ros-pkg/omnivelma
  - ================================================================================================================================== -->

<?xml version="1.0" ?>
<sdf version="1.6">
	<model name="velmwheel">

        <!-- ====================================================== Header ======================================================= -->

        <!-- Driver plugin for ROS integration -->
		<plugin name="velmwheel" filename="libvelmwheel_gazebo.so"/>
		<plugin name="velmwheel" filename="libmonokl_chain.so"/>
        <!-- Model's pose -->
		<pose>0 0 0.101 0 0 0</pose>
        <!-- Model type -->
		<static>false</static>
        
        <!-- ======================================================= Model ======================================================= -->
		
		<!-- Base block -->
        <link name="base">
			<pose>0 -0.081 0.019 0 0 0</pose>
			<collision name="base_collision">
				<geometry>
					<mesh>
						<uri>model://velmwheel/meshes/standard/collider/base.dae</uri>
					</mesh>
				</geometry>
			</collision>
			<inertial>
				<mass>60</mass>
				<inertia>
					<ixx>2.72</ixx>
					<ixy>0</ixy>
					<ixz>0</ixz>
					<iyy>2.94</iyy>
					<iyz>0</iyz>
					<izz>5.40</izz>
				</inertia>
			</inertial>
			<visual name="base_visual">
				<geometry>
					<mesh>
						<uri>model://velmwheel/meshes/standard/visual/base.dae</uri>
					</mesh>
				</geometry>
				<material>
					<script>
						<uri>model://velmwheel/materials/velmwheel.material</uri>
						<name>velmwheel/base</name>
					</script>
				</material>
			</visual>
		</link>
		
		<!-- Mobile front -->
        <link name="front">
			<pose>0 0.364 0.018 0 0 0</pose>
			<collision name="front_collision">
				<geometry>
					<mesh>
						<uri>model://velmwheel/meshes/standard/collider/front.dae</uri>
					</mesh>
				</geometry>
			</collision>
			<inertial>
				<mass>15</mass>
				<inertia>
					<ixx>0.0771</ixx>
					<ixy>0</ixy>
					<ixz>0</ixz>
					<iyy>0.7351</iyy>
					<iyz>0</iyz>
					<izz>0.7483</izz>
				</inertia>
			</inertial>
			<visual name="front_visual">
				<geometry>
					<mesh>
						<uri>model://velmwheel/meshes/standard/visual/front.dae</uri>
					</mesh>
				</geometry>
				<material>
					<script>
						<uri>model://velmwheel/materials/velmwheel.material</uri>
						<name>velmwheel/base</name>
					</script>
				</material>
			</visual>
		</link>
		
		<!-- LIDAR sensors (@todo calculate more precise positions)-->
        <include>
			<uri>model://monokl</uri>
			<name>monokl_r</name>
			<pose>0.3 0 0.176 0 0 -1.570796</pose>
		</include>
		<include>
			<uri>model://monokl</uri>
			<name>monokl_l</name>
			<pose>-0.3 0 0.176 0 0 1.570796</pose>
		</include>
		<joint name="monokl_r_fix" type="fixed">
			<parent>base</parent>
			<child>monokl_r::base</child>
		</joint>
		<joint name="monokl_l_fix" type="fixed">
			<parent>base</parent>
			<child>monokl_l::base</child>
		</joint>
		
		<!-- IMU sensor -->
        <include>
			<uri>model://wewucho</uri>
			<name>wewucho</name>
			<pose>0 0 0 0 0 0</pose>
		</include>
		<joint name="wewucho_fix" type="fixed">
			<parent>base</parent>
			<child>wewucho::base</child>
		</joint>
		
		<!-- Rear right wheel -->
        <link name="wheel_rr">
			<pose>0.38 -0.36 0 0 0 0</pose>
			<collision name="wheel_rr_collision">
				<geometry>
					<sphere>
						<radius>0.103</radius>
					</sphere>
				</geometry>
				<surface>
					<friction>
						<ode>
							<mu>infinity</mu>
							<mu2>0</mu2>
							<fdir1>0.707106781186548 0.707106781186548 0</fdir1>
							<slip1>0</slip1>
							<slip2>0</slip2>
						</ode>
						<bullet>
							<friction>1</friction>
							<friction2>0</friction2>
							<fdir1>0.707106781186548 0.707106781186548 0</fdir1>
							<rolling_friction>0</rolling_friction>
						</bullet>
					</friction>
				</surface>
			</collision>
			<inertial>
				<mass>1.6</mass>
				<inertia>
					<ixx>0.008</ixx>
					<ixy>0</ixy>
					<ixz>0</ixz>
					<iyy>0.004738485</iyy>
					<iyz>0</iyz>
					<izz>0.004738485</izz>
				</inertia>
			</inertial>
			<visual name="wheel_rr_middle">
				<geometry>
					<mesh>
						<uri>model://velmwheel/meshes/standard/visual/wheel.dae</uri>
					</mesh>
				</geometry>
				<material>
					<script>
						<uri>model://velmwheel/materials/velmwheel.material</uri>
						<name>velmwheel/wheels</name>
					</script>
				</material>
			</visual>
			<visual name="wheel_rr_rollers">
				<geometry>
					<mesh>
						<uri>model://velmwheel/meshes/standard/visual/wheel_l.dae</uri>
					</mesh>
				</geometry>
				<material>
					<script>
						<uri>model://velmwheel/materials/velmwheel.material</uri>
						<name>velmwheel/rollers</name>
					</script>
				</material>
			</visual>
		</link>
		
		<!-- Rear left wheel -->
        <link name="wheel_rl">
			<pose>-0.38 -0.36 0 0 0 0</pose>
			<collision name="wheel_rl_collision">
				<geometry>
					<sphere>
						<radius>0.103</radius>
					</sphere>
				</geometry>
				<surface>
					<friction>
						<ode>
							<mu>infinity</mu>
							<mu2>0</mu2>
							<fdir1>-0.707106781186548 0.707106781186548 0</fdir1>
							<slip1>0</slip1>
							<slip2>0</slip2>
						</ode>
						<bullet>
							<friction>1</friction>
							<friction2>0</friction2>
							<fdir1>-0.707106781186548 0.707106781186548 0</fdir1>
							<rolling_friction>0</rolling_friction>
						</bullet>
					</friction>
				</surface>
			</collision>
			<inertial>
				<mass>1.6</mass>
				<inertia>
					<ixx>0.008</ixx>
					<ixy>0</ixy>
					<ixz>0</ixz>
					<iyy>0.004738485</iyy>
					<iyz>0</iyz>
					<izz>0.004738485</izz>
				</inertia>
			</inertial>
			<visual name="wheel_rl_middle">
				<geometry>
					<mesh>
						<uri>model://velmwheel/meshes/standard/visual/wheel.dae</uri>
					</mesh>
				</geometry>
				<material>
					<script>
						<uri>model://velmwheel/materials/velmwheel.material</uri>
						<name>velmwheel/wheels</name>
					</script>
				</material>
			</visual>
			<visual name="wheel_rl_rollers">
				<geometry>
					<mesh>
						<uri>model://velmwheel/meshes/standard/visual/wheel_r.dae</uri>
					</mesh>
				</geometry>
				<material>
					<script>
						<uri>model://velmwheel/materials/velmwheel.material</uri>
						<name>velmwheel/rollers</name>
					</script>
				</material>
			</visual>
		</link>
		
		<!-- Front right wheel -->
        <link name="wheel_fr">
			<pose>0.38 0.36 0 0 0 0</pose>
			<collision name="wheel_fr_collision">
				<geometry>
					<sphere>
						<radius>0.103</radius>
					</sphere>
				</geometry>
				<surface>
					<friction>
						<ode>
							<mu>infinity</mu>
							<mu2>0</mu2>
							<fdir1>0.707106781186548 -0.707106781186548 0</fdir1>
							<slip1>0</slip1>
							<slip2>0</slip2>
						</ode>
						<bullet>
							<friction>1</friction>
							<friction2>0</friction2>
							<fdir1>0.707106781186548 -0.707106781186548 0</fdir1>
							<rolling_friction>0</rolling_friction>
						</bullet>
					</friction>
				</surface>
			</collision>
			<inertial>
				<mass>1.6</mass>
				<inertia>
					<ixx>0.008</ixx>
					<ixy>0</ixy>
					<ixz>0</ixz>
					<iyy>0.004738485</iyy>
					<iyz>0</iyz>
					<izz>0.004738485</izz>
				</inertia>
			</inertial>
			<visual name="wheel_fr_middle">
				<geometry>
					<mesh>
						<uri>model://velmwheel/meshes/standard/visual/wheel.dae</uri>
					</mesh>
				</geometry>
				<material>
					<script>
						<uri>model://velmwheel/materials/velmwheel.material</uri>
						<name>velmwheel/wheels</name>
					</script>
				</material>
			</visual>
			<visual name="wheel_fr_rollers">
				<geometry>
					<mesh>
						<uri>model://velmwheel/meshes/standard/visual/wheel_r.dae</uri>
					</mesh>
				</geometry>
				<material>
					<script>
						<uri>model://velmwheel/materials/velmwheel.material</uri>
						<name>velmwheel/rollers</name>
					</script>
				</material>
			</visual>
		</link>
		
		<!-- Front left wheel -->
        <link name="wheel_fl">
			<pose>-0.38 0.36 0 0 0 0</pose>
			<collision name="wheel_fl_collision">
				<geometry>
					<sphere>
						<radius>0.103</radius>
					</sphere>
				</geometry>
				<surface>
					<friction>
						<ode>
							<mu>infinity</mu>
							<mu2>0</mu2>
							<fdir1>-0.707106781186548 -0.707106781186548 0</fdir1>
							<slip1>0</slip1>
							<slip2>0</slip2>
						</ode>
						<bullet>
							<friction>1</friction>
							<friction2>0</friction2>
							<fdir1>-0.707106781186548 -0.707106781186548 0</fdir1>
							<rolling_friction>0</rolling_friction>
						</bullet>
					</friction>
				</surface>
			</collision>
			<inertial>
				<mass>1.6</mass>
				<inertia>
					<ixx>0.008</ixx>
					<ixy>0</ixy>
					<ixz>0</ixz>
					<iyy>0.004738485</iyy>
					<iyz>0</iyz>
					<izz>0.004738485</izz>
				</inertia>
			</inertial>
			<visual name="wheel_fl_middle">
				<geometry>
					<mesh>
						<uri>model://velmwheel/meshes/standard/visual/wheel.dae</uri>
					</mesh>
				</geometry>
				<material>
					<script>
						<uri>model://velmwheel/materials/velmwheel.material</uri>
						<name>velmwheel/wheels</name>
					</script>
				</material>
			</visual>
			<visual name="wheel_fl_rollers">
				<geometry>
					<mesh>
						<uri>model://velmwheel/meshes/standard/visual/wheel_l.dae</uri>
					</mesh>
				</geometry>
				<material>
					<script>
						<uri>model://velmwheel/materials/velmwheel.material</uri>
						<name>velmwheel/rollers</name>
					</script>
				</material>
			</visual>
		</link>
		
		<!-- Base-front joint -->
        <joint name="middle_joint" type="revolute">
			<parent>base</parent>
			<child>front</child>
			<!-- Pozycja względem dziecka -->
			<pose>0 -0.1 -0.018 0 0 0</pose>
			<axis>
				<xyz>0 1 0</xyz>
			</axis>
		</joint>
		
		<!-- Rear-right wheel's joint -->
        <joint name="motor_rr" type="revolute">
			<parent>base</parent>
			<child>wheel_rr</child>
			<pose>-0.036 0 0 0 0 0</pose>
			<axis>
				<xyz>-1 0 0</xyz>
			</axis><!--
			<physics>
				<ode>
					<velocity>10</velocity>
				</ode>
			</physics>-->
		</joint>
		
		<!-- Rear-left wheel's joint -->
        <joint name="motor_rl" type="revolute">
			<parent>base</parent>
			<child>wheel_rl</child>
			<pose>0.036 0 0 0 0 0</pose>
			<axis>
				<xyz>-1 0 0</xyz>
			</axis>
		</joint>
		
		<!-- Front-right wheel's joint -->
        <joint name="motor_fr" type="revolute">
			<parent>front</parent>
			<child>wheel_fr</child>
			<pose>-0.036 0 0 0 0 0</pose>
			<axis>
				<xyz>-1 0 0</xyz>
			</axis>
		</joint>
		
		<!-- Front-left wheel's joint -->
        <joint name="motor_fl" type="revolute">
			<parent>front</parent>
			<child>wheel_fl</child>
			<pose>0.036 0 0 0 0 0</pose>
			<axis>
				<xyz>-1 0 0</xyz>
			</axis>
		</joint>
	</model>
</sdf>

